src/storage/src/source/mod.rs

```diff
@@ -850,125 +965,553 @@ where
+                // downstream reclock operators.
+                let remap_trace = timestamper.remap_trace();
+                trace!(
+                    "remap({id}) {worker_id}/{worker_count}: \
+                    emitting initial remap_trace. \
+                    source_upper: {:?} \
+                    trace_updates: {:?}",
+                    global_source_upper,
+                    remap_trace
+                );
+
+                let mut remap_output = remap_output.activate();
+                let cap = cap_set.delayed(cap_set.first().unwrap());
+                let mut session = remap_output.session(&cap);
+                session.give(remap_trace);
+            }
+
+            while scheduler.notified().await {
+                if token_weak.upgrade().is_none() {
+                    // Make sure we don't accidentally mint new updates when this
+                    // source has been dropped. This way, we also make sure to not
+                    // react to spurious frontier advancements to `[]` that happen
+                    // when the input source operator is shutting down.
+                    return;
+                }
+
+                if !active_worker {
+                    // We simply loop because we cannot return here. Otherwise
+                    // the one active worker would not get frontier upgrades
+                    // anymore or other things could break.
+                    // WIP: Is this correct?
+                    continue;
+                }
+
+                input.for_each(|_cap, data| {
+                    data.swap(&mut buffer);
+
+                    for source_upper_summary in buffer.drain(..) {
+                        for (pid, offset) in source_upper_summary.source_upper {
+                            let previous_offset = global_source_upper.insert(pid, offset);
+                            if let Some(previous_offset) = previous_offset {
+                                assert!(previous_offset <= offset);
+                            }
                         }
                     }
-                    Event::Message(ts, message) => match message {
-                        Ok(message) => handle_message::<S>(
-                            message,
-                            &mut bytes_read,
-                            &cap_set,
-                            output,
-                            &mut metric_updates,
-                            ts,
-                        ),
-                        // TODO: make errors definite
-                        Err(e) => {
-                            output.session(&cap_set.delayed(&ts)).give(Err(SourceError {
-                                source_id: id,
-                                error: e.inner,
-                            }));
+                });
+
+                if let Err(wait_time) = timestamper.next_mint_timestamp() {
+                    activator.activate_after(wait_time);
```

Comment:
It's strange to see manual activations in an async operator. Does something break if we do `tokio::time::sleep(wait_time).await;` here?

Reply:
Will do! I did it because I didn't want to block reading of inputs while we wait.

Reply:
Ah: if we don't worry about not reading from the inputs, I can just remove the `if` entirely. The `timestamper.mint()` call just below this will `await` until it gets the next mint timestamp. WDYT?

Reply:
that makes sense to me, yeah!